### PR TITLE
TELCODOCS-1707: Add Performing upgrade GitOps

### PR DESCRIPTION
**Pull request by amolnar-gh**
Version(s): 4.16+

Issue: https://issues.redhat.com/browse/TELCODOCS-1707

Link to docs preview: https://77561--ocpdocs-pr.netlify.app/openshift-enterprise/latest/edge_computing/image_based_upgrade/ztp-image-based-upgrade.html

QE review:
- [ ] QE has approved this change.

Additional information: 
Dividing the original [PR](https://github.com/openshift/openshift-docs/pull/74025) to smaller chunks. This is PR#7 from [commit#8](https://github.com/openshift/openshift-docs/pull/74025/commits/87dd88f21e36e96c8d63cdcb19d30ae98daee286).

* [PR#1](https://github.com/openshift/openshift-docs/pull/76871)
* [PR#2](https://github.com/openshift/openshift-docs/pull/77078)
* [PR#3](https://github.com/openshift/openshift-docs/pull/77256)
* [PR#4](https://github.com/openshift/openshift-docs/pull/77279)
* [PR#5](https://github.com/openshift/openshift-docs/pull/77308)
* [PR#6](https://github.com/openshift/openshift-docs/pull/77379)
* [PR#7](https://github.com/openshift/openshift-docs/pull/77475)

The topicmap will be updated once the previous PRs are merged in, only relevant topicmap content is added. See thread in review channel on Slack.
Xrefs are pointing to non-existent sections at the moment, so they are commented out until added.

### CODE REVIEW

**Comment by amolnar-rh** [comment]
Please ignore the build error, that should be resolved after relevant PRs are merged and I rebase.

File: edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+include::modules/ztp-image-based-upgrade-talm-prep.adoc[leveloffset=+1]
+
+////
+[role="_additional-resources"]
+.Additional resources
+
+* xref:../../edge_computing/ztp-preparing-the-hub-cluster.adoc#ztp-preparing-the-ztp-git-repository-ver-ind_ztp-preparing-the-hub-cluster[Preparing the GitOps ZTP site configuration repository for version independence]
+
+* xref:../../edge_computing/image_based_upgrade/preparing_for_image_based_upgrade/ztp-image-based-upgrade-prep-resources.adoc[Preparing for the image-based upgrade for single-node OpenShift clusters]
```
[the remark this answers is not in the document]
Comment (tmalove) [reply edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc:26]: Same comment as the bot - missing anchor ID.

File: edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+* xref:../../edge_computing/image_based_upgrade/preparing_for_image_based_upgrade/cnf-image-based-upgrade-shared-container-image.adoc#ztp-image-based-upgrade-shared-container-directory_shared-container-directory[Configuring a shared container directory between ostree stateroots when using GitOps ZTP]
+
+* xref:../../backup_and_restore/application_backup_and_restore/installing/installing-oadp-ocs.adoc#oadp-about-backup-snapshot-locations_installing-oadp-ocs[About backup and snapshot locations and their secrets]
+
+* xref:../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/oadp-creating-backup-cr.adoc[Creating a Backup CR]
```
[the remark this answers is not in the document]
Comment (tmalove) [reply edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc:32]: Same comment as the bot - missing anchor ID.

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+. Apply the `Prep` policy:
++
+--
+[source,terminal]
+----
+$ oc apply -f cgu-ibu-prep.yml
+----
+
+If you provide `ConfigMap` objects for OADP resources and extra manifests, {lcao} validates the specified `ConfigMap` objects during the `Prep` stage.
+You may encounter the following issues:
```
Comment (tmalove) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:156]:
```suggestion
You might encounter the following issues: 
```
See ISG: https://www.ibm.com/docs/en/ibm-style?topic=word-usage#m

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+[source,terminal]
+----
+$ oc apply -f cgu-ibu-prep.yml
+----
+
+If you provide `ConfigMap` objects for OADP resources and extra manifests, {lcao} validates the specified `ConfigMap` objects during the `Prep` stage.
+You may encounter the following issues: 
+
+* validation warnings or errors if the {lcao} detects any issues with `extraManifests` 
+* validation errors if the {lcao} detects any issues with with `oadpContent`
```
Comment (tmalove) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:159]:
```suggestion
* validation errors if the {lcao} detects any issues with `oadpContent`
```
```suggestion
* validation errors if the {lcao} detects any issues with with `oadpContent`
```

**Comment by tmalove** [comment]
/remove-label peer-review-in-progress
/label peer-review-done

Also be sure to squash your commits before requesting a merge. Thanks.

**Review by tmalove — APPROVED** [review]

File: edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (amolnar-rh) [reply edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc:32]: These are commented out for now, I'll have a dedicated PR to add and resolve xref issues.

**Review by mpmaruthu — CHANGES_REQUESTED** [review]
few change needs to be updated otherwise looks good to me.

**Comment by mpmaruthu** [comment]
Telco RAN QE: Looks good to me.
/Peri M

File: edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc
Comment (jeana-redhat) [inline edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc:1]: Can replace GitOps ZTP with `{ztp}` throughout

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="ztp-image-based-upgrade-prep-gitops_{context}"]
+= Moving to the Prep stage of the image-based upgrade with {lcao} and {ztp}
+
+When you deploy the {lcao} on a cluster, an `ImageBasedUpgrade` CR is automatically created. You update this CR to specify the image repository of the seed image and to move through the different stages.
+
+.Prerequisites
+
+* Create policies and `ConfigMap` objects for resources used in the image-based upgrade. For more information, see _Creating ConfigMap objects for the image-based upgrade with {ztp}_
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:12]: I think you have the titles in quotes in the other ones I have looked at

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    - fileName: ibu/ImageBasedUpgrade.yaml
+      policyName: "finalize-stage-policy"
+      status:
+        conditions:
+          - reason: Idle
+            status: "True"
+            type: Idle
+----
+<1> The policy evaluation interval for compliant and non-compliant policies. Set them to `10s` to ensure that the policies status accurately reflects the current upgrade status.
+<2> Define the seed image, OpenShift Container Platform version, and pull secret for the upgrade in the Prep stage.
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:80]:
```suggestion
<2> Define the seed image, {product-title} version, and pull secret for the upgrade in the Prep stage.
```

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+<3> Define the OADP `ConfigMap` resources required for backup and restore.
+
+. Verify that the policies required for an image-based upgrade are created:
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:83]:
```suggestion
. Verify that the policies required for an image-based upgrade are created by running the following command:
```

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  - example-group-ibu-oadp-cm-policy
+  - example-group-ibu-prep-stage-policy
+  remediationStrategy:
+    canaries:
+      - spoke1
+    maxConcurrency: 1
+    timeout: 240
+----
+
+. Apply the `Prep` policy:
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:147]:
```suggestion
. Apply the `Prep` policy by running the following command:
```

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+[source,terminal]
+----
+$ oc apply -f cgu-ibu-prep.yml
+----
+
+If you provide `ConfigMap` objects for OADP resources and extra manifests, {lcao} validates the specified `ConfigMap` objects during the `Prep` stage.
+You might encounter the following issues: 
+
+* validation warnings or errors if the {lcao} detects any issues with `extraManifests` 
+* validation errors if the {lcao} detects any issues with `oadpContent`
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:159]:
```suggestion
* Validation warnings or errors if the {lcao} detects any issues with `extraManifests` 
* Validation errors if the {lcao} detects any issues with `oadpContent`
```

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+[...]
+----
+
+However, validation errors, such as adding `MachineConfig` or Operator manifests to extra manifests, cause the `Prep` stage to fail and block the `Upgrade` stage.
+
+When the validations pass, the cluster creates a new `ostree` stateroot, which involves pulling and unpacking the seed image, and running host level commands.
+Finally, all the required images are precached on the target cluster.
+--
+
+. Monitor the status and wait for the `cgu-ibu-prep` `ClusterGroupUpgrade` to report `Completed`.
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:180]:
```suggestion
. Monitor the status and wait for the `cgu-ibu-prep` `ClusterGroupUpgrade` to report `Completed` by running the following command:
```

File: modules/ztp-image-based-upgrade-talm-prep.adoc [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
++
+--
+[source,terminal]
+----
+$ oc get cgu -n default
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-prep.adoc:185]: would check with SMEs about `cgu` vs a long name as well (later)

File: modules/ztp-image-based-upgrade-talm-upgrade.adoc
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-upgrade.adoc:1]: {ztp} in here as well

File: modules/ztp-image-based-upgrade-talm-upgrade.adoc [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  remediationStrategy: 
+    canaries: 
+      - spoke1
+    maxConcurrency: 1 
+    timeout: 240
+----
+<1> Applies the `disable-auto-import` annotation to the managed cluster before starting the upgrade. This annotation ensures the automatic importing of managed cluster is disabled during the upgrade stage until the cluster is ready.
+<2> Removes the `disable-auto-import` annotation after the upgrade is complete.
+
+. Apply the `Upgrade` policy:
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-upgrade.adoc:51]:
```suggestion
. Apply the `Upgrade` policy by running the following command:
```

File: modules/ztp-image-based-upgrade-talm-upgrade.adoc [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
++
+[source,terminal]
+----
+$ oc apply -f cgu-ibu-upgrade.yml
+----
+
+.. Monitor the status and wait for the `cgu-ibu-upgrade` `ClusterGroupUpgrade` to report `Completed`.
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-upgrade.adoc:58]: Would un-nest this step `a` 
```suggestion
. Monitor the status by running the following command and wait for the `cgu-ibu-upgrade` `ClusterGroupUpgrade` to report `Completed`:
```

File: modules/ztp-image-based-upgrade-talm-upgrade.adoc [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    timeout: 240
+----
+
+[IMPORTANT]
+====
+Ensure that no other `ClusterGroupUpgrade` CRs are in progress because this causes {cgu-operator} to continuously reconcile them. Delete all `"In-Progress"` `ClusterGroupUpgrade` CRs before applying the `cgu-ibu-finalize.yaml`. 
+====
+--
+
+. Apply the policy:
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-upgrade.adoc:108]:
```suggestion
. Apply the policy by running the following command:
```

File: modules/ztp-image-based-upgrade-talm-upgrade.adoc [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
++
+[source,terminal]
+----
+$ oc apply -f cgu-ibu-finalize.yaml
+----
+
+.. Monitor the status and wait for the `cgu-ibu-finalize` `ClusterGroupUpgrade` to report `Completed`.
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-upgrade.adoc:115]: same here, would un-nest
```suggestion
. Monitor the status by running the following command and wait for the `cgu-ibu-finalize` `ClusterGroupUpgrade` to report `Completed`:
```

File: modules/ztp-image-based-upgrade-talm-rollback.adoc [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  managedPolicies: 
+  - example-group-ibu-rollback-stage-policy
+  remediationStrategy: 
+    canaries: 
+      - spoke1
+    maxConcurrency: 1 
+    timeout: 240
+----
+
+. Apply the `Rollback` policy:
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-rollback.adoc:68]:
```suggestion
. Apply the `Rollback` policy by running the following command:
```

File: modules/ztp-image-based-upgrade-talm-rollback.adoc [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+  managedPolicies: 
+  - example-group-ibu-finalize-stage-policy
+  remediationStrategy: 
+    canaries:
+      - spoke1
+    maxConcurrency: 1 
+    timeout: 240
+----
+
+. Apply the policy:
```
Comment (jeana-redhat) [inline modules/ztp-image-based-upgrade-talm-rollback.adoc:101]:
```suggestion
. Apply the policy by running the following command:
```

File: edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+* xref:../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/restoring-applications.adoc#oadp-creating-restore-cr_restoring-applications[Creating a Restore CR]
+////
+
+include::modules/ztp-image-based-upgrade-talm-upgrade.adoc[leveloffset=+1]
+
+include::modules/ztp-image-based-upgrade-talm-rollback.adoc[leveloffset=+1]
+
+include::modules/cnf-image-based-upgrade-troubleshooting.adoc[leveloffset=+1]
```
Comment (jeana-redhat) [inline edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc:39]: This one is not getting pulled into the assembly :thinking: 
image

**Review by jeana-redhat — COMMENTED** [review]
mostly minor stuff, but also the TS module is not getting rendered for some reason :thinking: 

/remove-label merge-review-in-progress
/remove-label merge-review-needed

File: edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc (hunk 18 above)
Comment (jeana-redhat) [reply edge_computing/image_based_upgrade/ztp-image-based-upgrade.adoc:39]: nvm this shows up now idk what the deal was

**Pull request merged**